File: frontend/src/services/userService.js

import api from './api';

const USERS_API_URL = '/users';

export const getAllUsers = async () => {
    const response = await api.get(USERS_API_URL);
    return response.data;
};

export const getUserById = async (id) => {
    const response = await api.get(`${USERS_API_URL}/${id}`);
    return response.data;
};

export const createUser = async (user) => {
    const response = await api.post(USERS_API_URL, user);
    return response.data;
};

export const updateUser = async (id, userDetails) => {
    const response = await api.put(`${USERS_API_URL}/${id}`, userDetails);
    return response.data;
};

export const deleteUser = async (id) => {
    await api.delete(`${USERS_API_URL}/${id}`);
};

export const existsByUsername = async (username) => {
    const response = await api.get(`${USERS_API_URL}/exists/username/${username}`);
    return response.data;
};

export const existsByEmail = async (email) => {
    const response = await api.get(`${USERS_API_URL}/exists/email/${email}`);
    return response.data;
};

File: backend/src/main/java/com/proj/sms/service/UserService.java
package com.proj.sms.service;

import com.proj.sms.interfaces.UserRepository;
import com.proj.sms.models.User;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import org.springframework.security.crypto.password.PasswordEncoder; // Add this
import org.springframework.stereotype.Service;

@Service
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public UserService(UserRepository userRepository, PasswordEncoder passwordEncoder) { // Modify constructor
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder; // Add this
    }


    // New method for login validation
    public Optional<User> validateUser(String username, String password) {
        return userRepository.findByUsername(username)
                .filter(user -> passwordEncoder.matches(password, user.getPassword()));
    }

    public User createUser(User user) {
        return userRepository.save(user);
    }

    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    public Optional<User> getUserById(Long id) {
        return userRepository.findById(id);
    }

    public Optional<User> getUserByUsername(String username) {
        return userRepository.findByUsername(username);
    }

    public Optional<User> getUserByEmail(String email) {
        return userRepository.findByEmail(email);
    }

    public User updateUser(Long id, User userDetails) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("User not found"));

        user.setUsername(userDetails.getUsername());
        user.setEmail(userDetails.getEmail());
        // Update other fields as needed

        return userRepository.save(user);
    }

    public void deleteUser(Long id) {
        userRepository.deleteById(id);
    }

    public boolean existsByUsername(String username) {
        return userRepository.existsByUsername(username);
    }

    public boolean existsByEmail(String email) {
        return userRepository.existsByEmail(email);
    }
}